Automatically handle back gesture in nested Navigators

justinmc opened the pull request:
There is a known problem with using nested Navigator widgets in Flutter. System back events will still go to the root Navigator instead of the nested one. The solution has been to use WillPopScope (and later PopScope/NavigatorPopHandler) to catch the system back and handle it manually. See the [docs on NavigatorPopHandler](https://main-api.flutter.dev/flutter/widgets/NavigatorPopHandler-class.html) for more on this problem and how we recommend solving it.

This PR is exploring the idea: why doesn't Flutter handle this complexity automatically for users? The navigation scenarios may be varied enough that Flutter can't solve this automatically without breaking some of them, but I hope to find out if that's true.

Fixes https://github.com/flutter/flutter/issues/145290
Fixes https://github.com/flutter/flutter/issues/153672

Discovered while working on this: https://github.com/flutter/flutter/issues/152578

### Multiple nested Navigators
I thought this would require more work, but it seems `maybePop` correctly doesn't pop the inactive Navigator.

<details>

<summary>Multiple nested Navigators example</summary>

```dart
import 'package:flutter/material.dart';

void main() async {
  runApp(const _MyApp());
}

class _MyApp extends StatelessWidget {
  const _MyApp();

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        useMaterial3: false,
        /*
        pageTransitionsTheme: const PageTransitionsTheme(
          builders: <TargetPlatform, PageTransitionsBuilder>{
            TargetPlatform.android: PredictiveBackPageTransitionsBuilder(),
          },
        ),
        */
      ),
      title: 'Flutter Demo',
      routes: {
        '/': (BuildContext context) {
          return _MyNestedNavigator();
        },
        'nav1leaf': (BuildContext context) {
          return const Nav1LeafPage();
        },
      },
    );
  }
}

class _MyNestedNavigator extends StatelessWidget {
  _MyNestedNavigator();

  final GlobalKey<NavigatorState> _nestedNavigatorKey = GlobalKey<NavigatorState>();

  @override
  Widget build(BuildContext context) {
    return Navigator(
      key: _nestedNavigatorKey,
      initialRoute: 'nav2home',
      onGenerateRoute: (RouteSettings settings) {
        return switch (settings.name) {
          'nav2home' => MaterialPageRoute(
            builder: (BuildContext context) => const MyHomePage(),
          ),
          'nav2leaf' => MaterialPageRoute(
            builder: (BuildContext context) => const Nav2LeafPage(),
          ),
          'nav3' => MaterialPageRoute(
            builder: (BuildContext context) => _MyNestedNestedNavigator(),
          ),
          _ => MaterialPageRoute(
            builder: (BuildContext context) => const Text('404'),
          ),
        };
      },
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blueAccent,
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Home page of Navigator 2'),
            const SizedBox(height: 16, width: double.infinity),
            ElevatedButton(
              child: const Text('Open nav 2 leaf'),
              onPressed: () => Navigator.of(context).pushNamed('nav2leaf'),
            ),
          ],
        ),
      ),
    );
  }
}

class Nav2LeafPage extends StatelessWidget {
  const Nav2LeafPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blue,
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          const Text('Leaf page of Navigator 2'),
          const SizedBox(height: 16, width: double.infinity),
          ElevatedButton(
            child: const Text('Open nav 3 home'),
            onPressed: () => Navigator.of(context).pushNamed('nav3'),
          ),
          const SizedBox(height: 16, width: double.infinity),
          ElevatedButton(
            child: const Text('Open nav 1 leaf'),
            onPressed: () => Navigator.of(context, rootNavigator: true).pushNamed('nav1leaf'),
          ),
        ],
      ),
    );
  }
}

class Nav1LeafPage extends StatelessWidget {
  const Nav1LeafPage({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Text('Leaf of Navigator 1'),
        ],
      ),
    );
  }
}

class _MyNestedNestedNavigator extends StatelessWidget {
  _MyNestedNestedNavigator();

  final GlobalKey<NavigatorState> _nestedNavigatorKey = GlobalKey<NavigatorState>();

  @override
  Widget build(BuildContext context) {
    return Navigator(
      key: _nestedNavigatorKey,
      initialRoute: 'nav3home',
      onGenerateRoute: (RouteSettings settings) {
        return switch (settings.name) {
          'nav3home' => MaterialPageRoute(
            builder: (BuildContext context) => const NestedNestedNavigatorHomePage(),
          ),
          'nav3leaf' => MaterialPageRoute(
            builder: (BuildContext context) => const Nav3LeafPage(),
          ),
          _ => MaterialPageRoute(
            builder: (BuildContext context) => const Text('404'),
          ),
        };
      },
    );
  }
}

class NestedNestedNavigatorHomePage extends StatelessWidget {
  const NestedNestedNavigatorHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.amber,
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Home page of Navigator 3'),
            const SizedBox(height: 16, width: double.infinity),
            ElevatedButton(
              child: const Text('Open nav 3 leaf'),
              onPressed: () => Navigator.of(context).pushNamed('nav3leaf'),
            ),
          ],
        ),
      ),
    );
  }
}

class Nav3LeafPage extends StatelessWidget {
  const Nav3LeafPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.amberAccent,
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          const Text('Leaf page of Navigator 3'),
          const SizedBox(height: 16, width: double.infinity),
          ElevatedButton(
            child: const Text('Open nav 1 leaf'),
            onPressed: () => Navigator.of(context, rootNavigator: true).pushNamed('nav1leaf'),
          ),
        ],
      ),
    );
  }
}
```

</details>

### TODO

 - [ ] More tests, if it looks like this PR is viable.

justinmc committed:
WIP exploring if I can automatically add a PopScope to nested Navigator widgets

justinmc committed:
Automatically catch back gestures in nested Navigators

justinmc committed:
Clean up

justinmc committed:
Fix CupertinoTabScaffold, which deactivates Navigator when its tab is active

The tabs are not managed by routes, they are just shown and hidden, so
this widget needs special custom handling of backs. This is a good
example of when we can't do this automatically.

justinmc committed:
handle to handles

justinmc committed:
Fix navigator tests by simplifying them.

They don't need to do PopScope themselves anymore.

justinmc committed:
Investigating predictive back not working when regular route transitions do work

Red herring, it only happens in complex interleaving scenarios, and that
is a bug in master too.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Docs for handlesBacksWhenNested

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc commented:
@maRci002 I'm interested in your thoughts here if you have a chance to take a look.

chunhtai commented on packages/flutter/lib/src/cupertino/tab_view.dart:
```diff
@@ -187,6 +187,10 @@ class _CupertinoTabViewState extends State<CupertinoTabView> {
+      // Manually use a NavigatorPopHandler to handle backs only on the active
+      // tab, instead of letting Navigator use its default back handling for
+      // nested Navigators.
+      handlesBacksWhenNested: false,
```
This is unfortunate. if there are multiple parallel navigators, what decide which one receives the pop?

justinmc replied:
See this test for how tabbed navigation is typically done here: https://github.com/flutter/flutter/blob/8d3d3ee4d078301fed75a1d6b984f0280a592fe4/packages/flutter/test/cupertino/tab_scaffold_test.dart#L1274

Each tab has its own nested Navigator wrapped by NavigatorPopHandler. Only the single active tab's NavigatorPopHandler is enabled though, so it's the only one that receives a call to onPop.

Or for the more generic case, where you have 2 sibling Navigators each wrapped with a PopScope, and both nested under a root Navigator? Both will be called. I think that's probably what we want. https://github.com/flutter/flutter/blob/8d3d3ee4d078301fed75a1d6b984f0280a592fe4/packages/flutter/lib/src/widgets/routes.dart#L1788

But for the even more generic case where you have two WidgetsBindingObservers listening for didPopRoute, only the one that was registered first will be called. I think that's very arbitrary and probably a mistake. It would probably solve a lot of the complexity with PopScope etc. if it was able to just call it on the current route or something.

https://github.com/flutter/flutter/blob/8d3d3ee4d078301fed75a1d6b984f0280a592fe4/packages/flutter/lib/src/widgets/binding.dart#L859

justinmc replied:
@chunhtai I wonder what you think about this handlesBacksWhenNested parameter versus scrapping this PR and adding NavigatorPopHandler to go_router?

maRci002 commented:
Here are some contexts for managing the back button in Flutter:

- **Using Back Button (without predictive back)**:
  - **WidgetsBinding.handlePopRoute** is invoked. For more details, see the [code here](https://github.com/flutter/flutter/blob/b0850beeb25f6d5b10426284f506557f66181b36/packages/flutter/lib/src/widgets/binding.dart#L833-L840).
  
  - **Non-router version**: 
    - `_WidgetsAppState.didPopRoute` is invoked. Refer to the [code here](https://github.com/flutter/flutter/blob/b0850beeb25f6d5b10426284f506557f66181b36/packages/flutter/lib/src/widgets/app.dart#L1537-L1552).
    - If you need to delegate the back button to a nested Navigator, use the techniques outlined in this [Flutter documentation on migrating from WillPopScope to NavigatorPopHandler for nested navigators](https://docs.flutter.dev/release/breaking-changes/android-predictive-back#migrating-from-willpopscope-to-navigatorpophandler-for-nested-navigators).
    
  - **Router version**:
    - `RootBackButtonDispatcher.didPopRoute` is triggered. For more information, view the [code here](https://github.com/flutter/flutter/blob/b0850beeb25f6d5b10426284f506557f66181b36/packages/flutter/lib/src/widgets/router.dart#L1073).
    - To delegate the back button to a nested Navigator, you should override `RouterDelegate.popRoute`. Depending on your AppState (_which should know which Navigator is active, e.g., when a `BottomNavigationBar` is used where each tab has its own Navigator_), find the appropriate `NavigatorState` by `GlobalKey`. See more [here](https://github.com/flutter/flutter/blob/b0850beeb25f6d5b10426284f506557f66181b36/packages/flutter/lib/src/widgets/router.dart#L1352).

- **Using Predictive Back**:
  - The following methods manage predictive back actions: `WidgetsBinding._handleStartBackGesture`, `_handleUpdateBackGestureProgress`, `_handleCommitBackGesture`, and `_handleCancelBackGesture`.
  - The method `_handleStartBackGesture` sets the `_backGestureObserver`, which handles the rest of the predictive back actions. See more [details here](https://github.com/flutter/flutter/blob/b0850beeb25f6d5b10426284f506557f66181b36/packages/flutter/lib/src/widgets/binding.dart#L845-L855).

Currently, we can delegate the back button to the appropriate `Navigator`, but we cannot delegate predictive back events to the appropriate `Navigator`. This pull request also aims to handle back button delegations?

I am unsure about the `canPop: !canPop(),` part. Does it mean that we intercept back button calls and allow predictive back anmimation only when the nested `Navigator` has none or a single page?

justinmc committed:
Update packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

justinmc commented:
@maRci002 What would that flag do?

This PR was not attempting to fix the problem with predictive back route transitions.

I'm just realizing that this PR does not fix the problem with ShellRoute either! Because go_router does not have a root Navigator. I need to reproduce that problem with just MaterialApp.router and no go_router so I fully understand it.

justinmc committed:
Example of nested Navigator.pages working with predictive back

justinmc committed:
Move the example up to the widgets library for better coverage

justinmc committed:
Reference example in docs

justinmc committed:
Merge branch 'master' into automatic-pop-scope

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -5681,6 +5704,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
+    // Navigator doesn't get all of them.
+    if (widget.handlesBacksWhenNested && Navigator.maybeOf(context, rootNavigator: true) != this) {
+      return PopScope(
+        canPop: !canPop(),
```
what would trigger the rebuild when canpop changes?

justinmc replied:
Updated to trigger a rebuild.

justinmc committed:
Analyzer fixes

justinmc committed:
WIP Needs state restoration

maRci002 commented:
> @maRci002 What would that flag do?

Basicly `WidgetsBindingObserver.handleStartBackGesture` would always return `false` hence other `Navigator`'s `Route` can be popped.

> This PR was not attempting to fix the problem with predictive back route transitions.

Oh, in that case, a more generic flag, such as `handlesBack`, would do the trick. `WidgetsBindingObserver.didPopRoute` could then check for this flag so basically only the 'current' Navigator will participate in the back events.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc commented:
@maRci002 I like your line of thought, but if I understand correctly, your approach won't work for multiple nested Navigators because WidgetsBinding iterates its observers in registration order. A middle Navigator would handle didPopRoute instead of the leaf one.

Consider the example in the description of this PR.

  * MaterialApp
    * Navigator1 (handlesBack=false)
      * _MyNestedNavigator
        * Navigator2 (handlesBack=true)
          * _MyNestedNestedNavigator
            * Navigator3 (handlesBack=true)

`_observers` will contain the three Navigators in this order (among other stuff):

 1. Navigator1
 2. Navigator2
 3. Navigator3

When it iterates them in handlePopRoute, it will call them in order:

 1. Navigator1.didPopRoute (false because handlesBack is false).
 2. Navigator2.didPopRoute (true, and it pops).

So Navigator2 will pop and Navigator3 won't get called, but we wanted Navigator3 to pop.

I think that WidgetsBinding.handlePopRoute is fundamentally flawed. Probably nested Navigators were not considered when it was written. Even if we changed it to iterate in reverse order, the order of the observers is not guaranteed, because it's possible that any of these Navigators could be disposed and rebuilt (and re-registered) without doing the same to the others.

I think it probably should have been written to notify all observers:

```dart
  Future<bool> handlePopRoute() async {
    bool handledPop = false;
    for (final WidgetsBindingObserver observer in List<WidgetsBindingObserver>.of(_observers)) {
      handledPop = await observer.didPopRoute() || handledPop;
    }
    if (handledPop) {
      return true;
    }
    SystemNavigator.pop();
    return false;
  }
```

If we do that plus your `handlesBack` parameter, then the user would be responsible for updating that parameter based on their current navigation stack. I'm still trying to figure out how I feel about that versus PopScope and everything...

maRci002 commented:
> If we do that plus your `handlesBack` parameter, then the user would be responsible for updating that parameter based on their current navigation stack.

Yes, this is key, and there won't be any breaking changes since `handlesBack` is set to true by default.

> A middle Navigator would handle didPopRoute instead of the leaf one.

User would be responsible for updating that parameter based on their current navigation stack.

> I think it probably should have been written to notify all observers:

In that case, multiple observers might perform a back event.

justinmc committed:
Fix the example

justinmc committed:
Look up the root navigator only ondidchangedependencies

justinmc committed:
Rebuild when canPop changes

justinmc committed:
No NavigatorPopHandler needed in our examples

Not sure what to do about the actual examples for NavigatorPopHandler. I
think I need one where we actually DO still need that widget. Not sure
what that looks like now though.

justinmc commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -5665,6 +5668,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
+
+    // If this is a nested Navigator, handle system backs here so that the root
+    // Navigator doesn't get all of them.
+    if (widget.handlesBacksWhenNested && Navigator.maybeOf(context, rootNavigator: true) != this) {
```
Done.

justinmc commented:
@maRci002  I don't see how your handlesBack approach will work without modifying handlePopRoute as I mentioned in my last comment. But if we modify handlePopRoute, then it will be a breaking change.

The main point of this PR is to fix nested Navigators in go_router (ShellRoute). I could also just open a PR in go_router that adds NavigatorPopHandler to it.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Missing license header

chunhtai commented:
This is superseded by https://github.com/flutter/flutter/pull/154718, @justinmc said they will close this pr if they manage to follow through the superseded one

justinmc commented:
On second thought https://github.com/flutter/flutter/pull/154718 is unrelated to the problem solved here. I think the decision to make is whether to do this automatically as in this PR, or to solve it specifically for go_router, maybe as in this PR: https://github.com/flutter/packages/pull/8162

justinmc commented:
I plan to revisit this PR in January: https://github.com/flutter/packages/pull/8162#issuecomment-2542575753

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3832,6 +3869,11 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
+    // If this is a nested Navigator, handle system backs so that the root
+    // Navigator doesn't get all of them.
+    _handlesBackGestures = widget.handlesBacksWhenNested
+        && Navigator.maybeOf(context, rootNavigator: true) != this;
```
This is unfortunate that rootNavigator needs to be set to true...

chunhtai replied:
can we do this in activate? `https://api.flutter.dev/flutter/widgets/State/activate.html`
and maybe do a assert make sure this won't change in didChangeDepedencies
I think this is the only possible place this can change, i.e. moving navigator around using globalkey.

chunhtai replied:
we also need to update _handlesBackGestures in didUpdateWidget in case the widget.handlesBacksWhenNested changes

justinmc replied:
> This is unfortunate that rootNavigator needs to be set to true...

For performance reasons you mean? The reason I do this is because I need to know if this Navigator is the root Navigator or not. I could do something like this but is it actually any better?

```dart
Element _getParentContext() {
  late final Element parentElement;
  context.visitAncestorElements((Element element) {
    parentElement = element;
    return false;
  });
  return parentElement;
}

bool isRootNavigator() {
  return Navigator.maybeOf(_getParentContext()) == null;
}
```

justinmc replied:
> can we do this in activate? `https://api.flutter.dev/flutter/widgets/State/activate.html` and maybe do a assert make sure this won't change in didChangeDepedencies I think this is the only possible place this can change, i.e. moving navigator around using globalkey.

Done in https://github.com/flutter/flutter/pull/152330/commits/5cb9fc457c713d6662d78f33c6ee3d4694e29465, thanks.

chunhtai reviewed (commented):
This approach LGTM. just left some comments

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Only rebuild for a canPop change when it matters.

This was breaking a test ('Able to pop all routes') that was rebuilding
with no routes, which triggered an assert.

justinmc committed:
Calculate isRootNavigator in activate instead of didChangeDeps, for performance

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc commented:
No, I need to clean this up. It's high risk but now that beta has been cut I'll get back to it.

reidbaker commented:
Part of the reason I ask is that we need this in the IO release and the beta that was just cut is the release that will be published as stable for io.

reidbaker commented:
Yes, this is (I think) the last part of our predictive back support which is being turned on by default in android 36 and announced at io.

justinmc commented:
Alright then let's do it! I will prioritize this and I'll keep a close eye out for problems after it lands.

In case we have to revert this, the other option would be to fix this directly in go_router.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
dart format

justinmc committed:
Fix a bug in didRemovePage.

Plus autoformatting, sorry for the bad diff.

justinmc committed:
dart format

justinmc committed:
No need for NavigatorPopHandle in showCupertinoSheet anymore

justinmc committed:
Remove now-unnecessary NavigatorPopHandler from some tests, keep for one that needs it

justinmc committed:
dart format

chunhtai reviewed (commented):
just some questions and some corner case

dkwingsmt commented:
Gentle ping @justinmc Are you planning to come back to this PR?

justinmc commented:
Yes, I'm planning to respond to code review and try to land this.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Remove print

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Test an unmigrated scenario

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -143,37 +143,30 @@ Future<T?> showCupertinoSheet<T>({
-      return NavigatorPopHandler(
-        onPopWithResult: (T? result) {
-          nestedNavigatorKey.currentState!.maybePop();
+      return Navigator(
```
It seems to not be. I've added a test for that in https://github.com/flutter/flutter/pull/152330/commits/f72897cd74dd9d7b9c629cceeed9956826208508.

I am a little worried, though. There could be some complicated situation someone is doing where things behave differently.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Analyzer

justinmc committed:
Explicitly don't handle backs in the WidgetsApp Navigator

Possibly avoiding weird situations where a nested Navigator is above the
WidgetsApp.

justinmc commented:
I have a fix for the Google test failures, and https://github.com/flutter/packages/pull/9856 is a fix for the go_router failures.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc commented:
The go_router fix has landed and rolled into Google, so hopefully those tests will pass now.

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc committed:
Consistent naming in example files

justinmc committed:
Use a getter to avoid an out of date value

justinmc committed:
Fix broken import from moving files

justinmc commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3784,6 +3823,8 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
     ServicesBinding.instance.accessibilityFocus.addListener(_recordLastFocus);
     _history.addListener(_handleHistoryChanged);
+
+    _isRootNavigator = _getIsRootNavigator();
```
Are you sure? The [docs](https://main-api.flutter.dev/flutter/widgets/State/activate.html) make it seem like it is needed:

 > The framework does not call this method the first time a [State] object
 > is inserted into the tree. Instead, the framework calls [initState] in
 > that situation.

justinmc committed:
Clean up by moving logic to setter

justinmc committed:
Fix docs reference to moved example

justinmc commented:
@chunhtai This is finally ready for re-review. I was worried that this PR would be too breaking, but now go_router’s breakages are fixed, the failing Google tests are very minimal (and I have a g3fix), and there are no customer test failures. I think I’m confident enough to try merging this. It should be much easier for users in the majority of cases, where they don’t want to think about handling predictive back gestures manually for their nested Navigators.

justinmc committed:
Getter/property/setter

justinmc committed:
Future proof in case called not during build

justinmc commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3674,6 +3698,27 @@ class _History extends Iterable<_RouteEntry> with ChangeNotifier {
+
+  bool _lastCanPopCached = false; // Because history is empty to start.
+  bool get _lastCanPop => _lastCanPopCached;
+  set _lastCanPop(bool value) {
```
Good point. I'll add a comment here to help explain lastCanPop in the meantime.

justinmc committed:
Comments on lastCanPop

justinmc commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1648,6 +1650,28 @@ class Navigator extends StatefulWidget {
+  ///  * [PopScope], which provides even more control over system back behavior.
+  ///  * [CupertinoTabView], which sets this parameter to false in order to
+  ///    provide custom back handling for its nested [Navigator]s.
+  final bool handlesBacksWhenNested;
```
That's what `true` does currently, see _isRootNavigator and _getIsRootNavigator.

I think there is really no third state to be handled. When the Navigator is not root, then it doesn't need a PopScope widget, it will handle back gestures anyway. So an unqualified `handlesBacks` set to true would mean the same thing as handlesBacksWhenNested is true. And it's not really possible to not handle backs at all for a root Navigator.

Maybe the "WhenNested" part of the name makes this confusing.

chunhtai replied:
sgtm

chunhtai reviewed (dismissed):
LGTM, just one suggestion

justinmc committed:
Merge branch 'master' into automatic-pop-scope

justinmc commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -1679,6 +1686,7 @@ class _WidgetsAppState extends State<WidgetsApp> with WidgetsBindingObserver {
         debugLabel: 'Navigator Scope',
         autofocus: true,
         child: Navigator(
+          handlesBacksWhenNested: false,
```
I think I did this specifically to work around a weird setup like that in the Google tests. I will try it and see if it passes, maybe that's out of date or I'm misremembering.

justinmc committed:
Shouldn't be needed, but is it working around a Google failure?

justinmc committed:
Merge branch 'master' into automatic-pop-scope

Piinks commented:
I've marked this PR as not ready to port to flutter/packages yet.
We'll provide instructions to move this change over to material_ui/cupertino_ui once ready to receive PRs. Thank you!

The changes not in the material or cupertino libraries will remain here.

Piinks commented:
Advice for PR splitting: Since this touches material/cupertino along with other parts of flutter/flutter, it will need to be split up.

If the material/cupertino changes are dependent on the other non-material/cupertino changes in this PR, then the material/cupertino changes will need to wait for them to roll to stable before they can be ported over to material_ui/cupertino_ui in flutter/packages (which uses the stable branch).
- If that is the case here, consider opening a new PR with the non-material/cupertino changes to move forward with so that waiting period can be reduced. Up to you!

reidbaker commented:
@justinmc what is the status of this pr do you want review?

Piinks commented:
When ready, material_ui and cupertino_ui are now accepting contributions. The portions of this PR that affect the material and cupertino libraries can be ported over to flutter/packages. Guidance and instructions for easily porting over can be found in https://github.com/flutter/flutter/issues/188444

justinmc commented:
I'll close this since it does have a few places where it touches Cupertino.
